clippy_lints/src/single_field_patterns.rs

```diff
@@ -0,0 +1,264 @@
+#![allow(rustc::usage_of_ty_tykind)]
+
+use clippy_utils::{
+    diagnostics::{multispan_sugg_with_applicability, span_lint_and_then},
+    higher::IfLetOrMatch,
+    higher::WhileLet,
+    source::snippet_opt,
+};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, Local, MatchSource, Pat, PatKind, Stmt, StmtKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{symbol::Ident, Span};
+use std::iter::once;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///  Checks for patterns that only use a single field when they could directly access the field.
+    ///
+    /// ### Why is this bad?
+    ///  It requires more text and more information than directly accessing the field.
+    ///
+    /// ### Example
+    /// ```rust
+    /// # struct Struct {
+    /// #     field1: Option<i32>,
+    /// #     field2: Option<i32>,
+    /// # }
+    /// # fn foo(struct1: Struct) {
+    /// // bad:
+    /// match struct1 {
+    ///     Struct { field1: Some(n), .. } if n >= 50 => {},
+    ///     Struct { field1: None, .. } => {},
+    ///     _ => {},
+    /// }
+    /// // better:
+    /// match struct1.field1 {
+    ///     Some(n) if n >= 50 => {},
+    ///     None => {},
+    ///     _ => {},
+    /// }
+    /// # }
+    /// ```
+    #[clippy::version = "1.59.0"]
+    pub SINGLE_FIELD_PATTERNS,
+    style,
+    "single-field patterns"
+}
+declare_lint_pass!(SingleFieldPatterns => [SINGLE_FIELD_PATTERNS]);
+
+/// This represents 0 or 1 fields being used. Where more may be used, I use Option<SingleField>
+///   where None represents the absence of a lint
+#[derive(Debug, Clone, Copy)]
+enum SingleField {
+    Id { id: Ident, pattern: Span },
+    Unused, // The name "SingleField" is a lie but idk what's better. "AtMostOneField"?
+}
+
+impl PartialEq for SingleField {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (SingleField::Id { id: id1, .. }, SingleField::Id { id: id2, .. }) => id1 == id2,
+            (SingleField::Unused, SingleField::Unused) => true,
+            _ => false,
+        }
+    }
+}
+
+fn get_sf<'a>(mut iter: impl Iterator<Item = (Ident, &'a Pat<'a>)>) -> Option<SingleField> {
+    let one = iter.by_ref().find(|(_, pat)| !matches!(pat.kind, PatKind::Wild));
+    match one {
+        Some((id, pat)) => {
+            if pat.span.from_expansion() {
+                return None;
+            }
+            if iter.all(|(_, other)| matches!(other.kind, PatKind::Wild)) {
+                Some(SingleField::Id { id, pattern: pat.span })
+            } else {
+                None
+            }
+        },
+        None => Some(SingleField::Unused),
+    }
+}
+
+fn struct_sf(pat: &PatKind<'_>) -> Option<SingleField> {
+    match *pat {
+        PatKind::Struct(_, pats, _) => get_sf(pats.iter().map(|field| (field.ident, field.pat))),
+        _ => None,
+    }
+}
```

Comment:
This should then also go into the impl block. Since the get_sf fn is only used in this function, there's probably no reason to separate those.